CA integration through Envoy SDS

maxlambrecht opened the pull request:
**Please provide a description of this PR:**

This PR includes the following changes: 

1. Make `istio-agent` check the presence of a Unix Domain Socket on the fixed path `/var/run/secrets/workload-identity/socket`:

           1. In the case where the socket exists, istio-agent does not start its own SDS Server.
           2. In the case where the socket does not exist, istio-agent checks whether certificate files on the fixed path folder (/var/run/secrets/workload-identity) exists.  If the certificate files exist, istio-agent starts its own SDS Server, listening on the fixed socket path (/var/run/secrets/workload-identity/socket); to serve those certificate files, keeping file  watchers on them. Envoy proxy connects to the SDS Server run by istio-agent through the fixed socket path and gets the cryptographic materials of the certificate files served by the SDS API.
           3. In case istio-agent does not find either the socket, or the certificate files in the fixed path;  it starts its own SDS Server using a `caClient` to connect to istiod to get the cryptographic materials (See CA Flow).

 Added new tests for istio-agent covering the scenarios where the socket is detected and where the certificates are detected.

2.  Added a volume and volumeMount for the fixed path to the chart manifests. 

3. Updated all the affected testdata files. 

4. Update agent README with the details of the new CA integration approach.

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [x] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [X] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

Commit:
Update build_push_update_images to use trivy instead of imagescanner (#37173)

istio-testing committed:
Automator: update common-files@master in istio/istio@master (#37186)

Commit:
Update release-builder dep + others (-proxy) (#37179)

* Update release-builder dep + others (-proxy)

* Rerun to get new repo updates

* Pick up latest release-builder

Commit:
respect prometheus annotations (#36959)

* respect prometheus annotations

* update comment

* refactor applyPrometheusMerge

* fix lint

* add release notes

* fix sets

* Update releasenotes/notes/36959.yaml

Co-authored-by: Eric Van Norman <ericvn@us.ibm.com>

Co-authored-by: Eric Van Norman <ericvn@us.ibm.com>

howardjohn committed:
tf: speed up telemetry tests (#37158)

* tf: speed up telemetry tests

* fix

* gen

* fix lint

istio-testing committed:
Automator: update istio/client-go@master dependency in istio/istio@master (#37187)

arkodg committed:
Fix IOP merge for ingressgategateway label field (#37191)

* Fix IOP merge for ingressgategateway label field

Fixes: https://github.com/istio/istio/issues/37190

Signed-off-by: Arko Dasgupta <arko@tetrate.io>

* fix testdata output

Signed-off-by: Arko Dasgupta <arko@tetrate.io>

kebe7jun committed:
Fix ValidatingWebhookConfiguration naming in precheck (#36548)

kebe7jun committed:
Remove useless rootca-compare alias (#36550)

bianpengyuan committed:
Remove liveness prober from CNI Daemonset. (#37018)

maxlambrecht committed:
CA integration through SDS: agent implementation. Unit tests.

Signed-off-by: Max Lambrecht <maxlambrecht@gmail.com>

maxlambrecht committed:
CA integration through SDS: agent implementation. Unit tests.

Signed-off-by: Max Lambrecht <maxlambrecht@gmail.com>

maxlambrecht committed:
CA integration through SDS: adding volume and volumeMount to chart manifests

Signed-off-by: Max Lambrecht <maxlambrecht@gmail.com>

maxlambrecht committed:
CA integration through SDS: adding volume and volumeMount to chart manifests

Signed-off-by: Max Lambrecht <maxlambrecht@gmail.com>

maxlambrecht committed:
CA integration through SDS: add workload-identity volume to testdata files

Signed-off-by: Max Lambrecht <maxlambrecht@gmail.com>

maxlambrecht committed:
CA integration through SDS: add workload-identity volume to testdata files

Signed-off-by: Max Lambrecht <maxlambrecht@gmail.com>

maxlambrecht committed:
CA integration through SDS: adding details to agent README.

Signed-off-by: Max Lambrecht <maxlambrecht@gmail.com>

maxlambrecht committed:
CA integration through SDS: adding details to agent README.

Signed-off-by: Max Lambrecht <maxlambrecht@gmail.com>

mgbcaio commented on pkg/config/constants/constants.go:
```diff
@@ -42,6 +42,21 @@ const (
+	// WorkloadIdentitySocketPath is the path to the Unix Domain Socket for SDS.
+	WorkloadIdentitySocketPath = WorkloadIdentityPath + "socket"
+
+	// WorkloadIdentityCertChainPath is path to a workload certificate chain file.
```
```suggestion
	// WorkloadIdentityCertChainPath is the path to a workload certificate chain file.
```

mgbcaio commented on pkg/config/constants/constants.go:
```diff
@@ -42,6 +42,21 @@ const (
+	// WorkloadIdentityCertChainPath is path to a workload certificate chain file.
+	WorkloadIdentityCertChainPath = WorkloadIdentityPath + "cert-chain.pem"
+
+	// WorkloadIdentityKeyPath is path to a workload key file.
```
```suggestion
	// WorkloadIdentityKeyPath is the path to a workload key file.
```

mgbcaio commented on pkg/config/constants/constants.go:
```diff
@@ -42,6 +42,21 @@ const (
+	// WorkloadIdentityKeyPath is path to a workload key file.
+	WorkloadIdentityKeyPath = WorkloadIdentityPath + "key.pem"
+
+	// WorkloadIdentityRootCertPath is path to a workload root certificate file.
```
```suggestion
	// WorkloadIdentityRootCertPath is the path to a workload root certificate file.
```

mgbcaio commented on tools/packaging/common/envoy_bootstrap.json:
```diff
@@ -245,7 +245,7 @@
                 "address":{
                   "pipe": {
-                    "path": "{{ .config.ConfigPath }}/SDS"
+                    "path": "./var/run/secrets/workload-identity/socket"
```
Do we need to parametrize this path or it will always be in the same place all the time ?
I know that in the README it says it will be a fixed path, just wondering if this could be parametrized to have this configured easily in any scenario it might come up, and not require a code change to actually change the path

maxlambrecht replied:
It will be a fixed path, so Envoy always tries to connect to a socket on that path to get its certificates, and istio-agent always detects the presence or starts its own SDS server on the same path. This was agreed upon on a follow-up discussion with the istio community.

mgbcaio commented on pkg/config/constants/constants.go:
```diff
@@ -42,6 +42,21 @@ const (
 	// ConfigPathDir config directory for storing envoy json config files.
 	ConfigPathDir = "./etc/istio/proxy"
 
+	// WorkloadIdentityPath is the path to the folder where workload identity materials and the SDS socket are placed.
```
```suggestion
	// WorkloadIdentityPath is the path to the folder where the workload identity materials and the SDS socket are placed.
```

mgbcaio commented on pkg/istio-agent/agent.go:
```diff
@@ -403,14 +403,36 @@ func (a *Agent) Run(ctx context.Context) (func(), error) {
+				return nil, fmt.Errorf("failed to start workload secret manager %v", err)
+			}
+		} else {
+			log.Info("workload identity cert files not found, create secret manager with caClient")
```
```suggestion
			log.Info("workload identity cert files not found, creating secret manager with the caClient")
```

mgbcaio commented on pkg/istio-agent/agent.go:
```diff
@@ -403,14 +403,36 @@ func (a *Agent) Run(ctx context.Context) (func(), error) {
+		certChainExists := fileExists(constants.WorkloadIdentityCertChainPath)
+		keyExists := fileExists(constants.WorkloadIdentityKeyPath)
+		if rootCertExists && certChainExists && keyExists {
+			log.Info("workload identity cert files detected, creating secret manager without caClient")
```
```suggestion
			log.Info("workload identity cert files detected, creating secret manager without the caClient")
```

mgbcaio commented on pkg/istio-agent/README.md:
```diff
@@ -14,6 +14,20 @@ requests to the agent, which will be forwarded to the configured discovery serve
+istio-agent checks the presence of a socket file on the fixed path `/var/run/secrets/workload-identity/socket` 
+
+1. In the case where the socket exists, istio-agent does not start its own SDS Server.
+2. In the case where the socket does not exist, istio-agent checks whether certificate files on the fixed path folder (/var/run/secrets/workload-identity) exists. 
```
```suggestion
2. In the case where the socket does not exist, istio-agent checks whether the certificate files are present on the fixed path folder (/var/run/secrets/workload-identity) exists. 
```

mgbcaio commented on pkg/istio-agent/README.md:
```diff
@@ -14,6 +14,20 @@ requests to the agent, which will be forwarded to the configured discovery serve
+1. In the case where the socket exists, istio-agent does not start its own SDS Server.
+2. In the case where the socket does not exist, istio-agent checks whether certificate files on the fixed path folder (/var/run/secrets/workload-identity) exists. 
+   If the certificate files exist, istio-agent starts its own SDS Server, listening on the fixed socket path (/var/run/secrets/workload-identity/socket); 
+   to serve those certificate files, keeping file  watchers on them. Envoy proxy connects to the SDS Server run by istio-agent 
```
```suggestion
   to serve those certificate files, keeping file watchers on them. Envoy proxy connects to the SDS Server run by istio-agent 
```

mgbcaio reviewed (changes requested):
Nice job guys! really clean code and nice to read. 

Left some nit-pick suggestions, feel free to ignore if you think it's not needed.

🥇

Commit:
Minor fixes on comments and docs.

Co-authored-by: Caio Milfont <93996310+mgbcaio@users.noreply.github.com>

Commit:
Minor fixes on comments and docs.

Co-authored-by: Caio Milfont <93996310+mgbcaio@users.noreply.github.com>

yskopets committed:
vm support: fix incorrect behaviour of the `hydrateWorkloadInstance()` (#37206)

* ci: add unit test for `hydrateWorkloadInstance`

Signed-off-by: Yaroslav Skopets <yaroslav@tetrate.io>

* vm support: fix hydrateWorkloadInstance()

Signed-off-by: Yaroslav Skopets <yaroslav@tetrate.io>

* fix linting issues

Signed-off-by: Yaroslav Skopets <yaroslav@tetrate.io>

jonkerj committed:
allow node selection parameters in operator chart (#36714)

All credits go to @matheusfm. This is a rebase of his commit/PR, since
his was closed before anyone took a chance to look at it.

Signed-off-by: Jorik Jonker <jorik@kippendief.biz>

hzxuzhonghu committed:
Credential controller: make use of generic event handler wrapper (#37075)

* Pass cb when instantialize Credential controller

* Fix comments

* goimports

* revert changes to handler registration

* remove comment

* goimports

mKeRix committed:
Fix IPv6 detection for EKS (#37164)

* Fix IPv6 detection for EKS

EKS clusters on IPv6 map a host-local NAT interface to each pod for IPv4
egress. For that the pod will be assigned an IP from the link-local
reserved CIDR range that is unique only to the host. For proper IPv6
detection we need to exclude these IPs from the discovery mechanisms.

* Add release note

* Exclude link-local multicast addresses

howardjohn committed:
Treat NaN as equal for fuzzer (#37212)

nmittler committed:
Manually updating istio/api (#37213)

Attempting to resolve issues in https://github.com/istio/istio/pull/37207

Co-authored-by: istio-testing <istio-testing-bot@google.com>

jwendell committed:
Only setup the credentials if there is one (#37219)

Otherwise the credential helpers will be installed globally and this
might mess with pulls from `gcr.io`.

howardjohn committed:
Revert "Only setup the credentials if there is one (#37219)" (#37222)

This reverts commit d00b521f26eef3033f40ae54534eb721b2cbd456.

aryan16 committed:
added tests for fromHeader and fromParams in request authentication (#37061)

* added tests for fromHeader and fromParams in request authentication

* fixed lint issues

* added tests for set of tokens in query params

Co-authored-by: Aryan Gupta <garyan@google.com>

istio-testing committed:
Automator: update istio/client-go@master dependency in istio/istio@master (#37208)

howardjohn committed:
Drop invalid filters instead of panic (#37214)

This only impacts tests

howardjohn committed:
gateway: allow ServiceImport as backendRef (#37216)

This has two behaviors:
* if clusterset.local is enabled, we change it to forward to
  clusterset.local cluster
* else, we treat it the same as Service. This is still useful, so folks
  can have portable routes between implementations

kebe7jun committed:
Make helloworld app listen both ipv6 and ipv4 addresses (#37224)

ramaraochavali committed:
send unhealthy endpoints to Envoy (take 2) (#37122)

* send unready endpoints as unhealthy

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* comment

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* set ready

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* set more ready and debug log

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* move to enum

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* lint

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* do not push if only not ready endpoints are there for service

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* fix ut

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* lint

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* add unit test

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* refactor, add feature flag and release notes

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* fix lint and release notes

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* lint

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* fix comments

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* address comments

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* minor change

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* move inside lock

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* update instancesByport to use unhealthy addresses

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* fix ut

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* fix lint

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

hanxiaop committed:
fix x authz description (#37068)

Commit:
minor refactor `util/sets` (#37230)

kerthcet committed:
fix typo and do some refactoring work (#37047)

Signed-off-by: kerthcet <kerthcet@gmail.com>

kerthcet committed:
break down NewStatusVerifier (#37180)

Signed-off-by: kerthcet <kerthcet@gmail.com>

howardjohn committed:
Skip flaky TestWait (#37210)

See https://github.com/istio/istio/issues/29315. This is flaky and known
to be a real bug in an experimental feature. Disable until fixed.

istio-testing committed:
Automator: update istio/api@master dependency in istio/istio@master (#37233)

jufantozzi commented on pkg/istio-agent/README.md:
```diff
@@ -14,6 +14,20 @@ requests to the agent, which will be forwarded to the configured discovery serve
+   If the certificate files exist, istio-agent starts its own SDS Server, listening on the fixed socket path (/var/run/secrets/workload-identity/socket); 
+   to serve those certificate files, keeping file watchers on them. Envoy proxy connects to the SDS Server run by istio-agent 
+   through the fixed socket path and gets the cryptographic materials of the certificate files served by the SDS API.
+3. In case istio-agent does not find either the socket, or the certificate files in the fixed path; 
```
```suggestion
3. In case istio-agent does not find either the socket, or the certificate files in the fixed path
```

jufantozzi commented on pkg/istio-agent/agent.go:
```diff
@@ -403,14 +403,36 @@ func (a *Agent) Run(ctx context.Context) (func(), error) {
-	a.secretCache, err = a.newSecretManager()
-	if err != nil {
-		return nil, fmt.Errorf("failed to start workload secret manager %v", err)
+	sdsSocketExists := socketFileExists(constants.WorkloadIdentitySocketPath)
```
I've been looking a lot at this section and wondering if they're not gonna complain about `Agent.Run()` getting too large, or ask us to flatten the logic (to avoid stacking if/elses inside of each other) but also having a hard time figuring out how improve on these matters 😅

alexandrealvino commented on pkg/istio-agent/README.md:
```diff
@@ -14,6 +14,20 @@ requests to the agent, which will be forwarded to the configured discovery serve
+   to serve those certificate files, keeping file watchers on them. Envoy proxy connects to the SDS Server run by istio-agent 
+   through the fixed socket path and gets the cryptographic materials of the certificate files served by the SDS API.
+3. In case istio-agent does not find either the socket, or the certificate files in the fixed path; 
+   it starts its own SDS Server using a `caClient` to connect to istiod to get the cryptographic materials (See CA Flow).
```
```suggestion
   it starts its own SDS Server using a `caClient` to connect to istiod to get the cryptographic materials (See Default CA Flow).
```

jufantozzi reviewed (approved):
Looking good! just left one observation

istio-testing committed:
Automator: update istio/client-go@master dependency in istio/istio@master (#37234)

alexandrealvino commented on pkg/istio-agent/agent.go:
```diff
@@ -403,14 +403,36 @@ func (a *Agent) Run(ctx context.Context) (func(), error) {
-		return nil, fmt.Errorf("failed to start workload secret manager %v", err)
+	sdsSocketExists := socketFileExists(constants.WorkloadIdentitySocketPath)
+	if sdsSocketExists {
+		log.Info("master branch: SDS socket detected, don't start SDS Server")
+	} else {
+		log.Info("master branch: SDS socket not detected, starting owned SDS Server")
```
Why `log.Info("master branch:`?

jufantozzi replied:
I was wondering the same

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#37004)

howardjohn committed:
Use stable name for dumping (#37237)

Currently we use the real cluster name. This is hard to debug, as we
will see a test like `from_primary-1`, but then folder is
`foo-cluster-1236`. We may not know which is which.

howardjohn committed:
Workaround centos repo removal (#37238)

Cherrypick https://github.com/istio/istio/pull/37090

nmittler committed:
Switch default LB algorithm to LEAST_REQUEST (#37220)

Also adding an environment variable to revert the default globally.

istio-testing committed:
Update BASE_VERSION to master-2022-02-09T20-27-12 (#37242)

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#37252)

Commit:
add insecure oci support (#36903)

* add insecure oci support

* Update pkg/wasm/imagefetcher.go

Co-authored-by: Pengyuan Bian <bianpengyuan@google.com>

* fix nit

Co-authored-by: Pengyuan Bian <bianpengyuan@google.com>

istio-testing committed:
Automator: update istio/client-go@master dependency in istio/istio@master (#37255)

yskopets committed:
vm support: support multiple WorkloadEntry(s) with the same IP (#37129)

* vm support: support multiple WorkloadEntry(s) with the same IP

Signed-off-by: Yaroslav Skopets <yaroslav@tetrate.io>

* add unit tests

Signed-off-by: Yaroslav Skopets <yaroslav@tetrate.io>

* fix linting issues

Signed-off-by: Yaroslav Skopets <yaroslav@tetrate.io>

* refactoring: move `workloadinstances` package

Signed-off-by: Yaroslav Skopets <yaroslav@tetrate.io>

* apply similar changes to `serviceregistry/serviceentry`

Signed-off-by: Yaroslav Skopets <yaroslav@tetrate.io>

* fix lint issues

Signed-off-by: Yaroslav Skopets <yaroslav@tetrate.io>

* refactor out `serviceregistry/serviceentry` changes into a separate PR

Signed-off-by: Yaroslav Skopets <yaroslav@tetrate.io>

* refactor out `serviceregistry/serviceentry` changes into a separate PR

Signed-off-by: Yaroslav Skopets <yaroslav@tetrate.io>

* fix unit tests

Signed-off-by: Yaroslav Skopets <yaroslav@tetrate.io>

* add more unit tests

Signed-off-by: Yaroslav Skopets <yaroslav@tetrate.io>

* code review: avoid creating a new set

Signed-off-by: Yaroslav Skopets <yaroslav@tetrate.io>

* code review: introduce `fx.WaitOrFail()` and `fx.WaitForDurationOrFail()`

Signed-off-by: Yaroslav Skopets <yaroslav@tetrate.io>

maxlambrecht committed:
CA integration through SDS: amendments in logging.

Signed-off-by: Max Lambrecht <maxlambrecht@gmail.com>

maxlambrecht committed:
CA integration through SDS: amendments in logging.

Signed-off-by: Max Lambrecht <maxlambrecht@gmail.com>

Commit:
CA integration through SDS: amendments in docs

Co-authored-by: Juliano Fantozzi <juliano.fantozzi@gmail.com>
Co-authored-by: Alexandre Alves Alvino <alexandre.alvino@stone.com.br>

Commit:
CA integration through SDS: amendments in docs

Co-authored-by: Juliano Fantozzi <juliano.fantozzi@gmail.com>
Co-authored-by: Alexandre Alves Alvino <alexandre.alvino@stone.com.br>

maxlambrecht committed:
CA integration through SDS: adding release notes.

Signed-off-by: Max Lambrecht <maxlambrecht@gmail.com>

maxlambrecht committed:
CA integration through SDS: adding release notes.

Signed-off-by: Max Lambrecht <maxlambrecht@gmail.com>

maxlambrecht committed:
Merge branch 'sds-integration' of https://github.com/maxlambrecht/istio into sds-integration

maxlambrecht committed:
CA integration through SDS: additions to the testdata.

Signed-off-by: Max Lambrecht <maxlambrecht@gmail.com>

maxlambrecht committed:
Merge branch 'sds-integration' of https://github.com/maxlambrecht/istio into sds-integration

maxlambrecht commented:
Thanks for all the comments, I'm closing this PR and I'm sending a new one including several changes that were suggested in the proposal document.
